toadr3/serializer.py

```diff
@@ -0,0 +1,252 @@
+            case Report():
+                result = self.serialize_report(obj)
+            case ReportDescriptor():
+                result = self.serialize_report_descriptor(obj)
+            case ReportPayloadDescriptor():
+                result = self.serialize_report_payload_descriptor(obj)
+            case Event():
+                result = self.serialize_event(obj)
+            case EventPayloadDescriptor():
+                result = self.serialize_event_payload_descriptor(obj)
+            case _:
+                result = super().default(obj)
+        return result
+
+    def serialize_interval_period(self, obj: IntervalPeriod) -> dict[str, Any]:
+        """Serialize an IntervalPeriod object to JSON encodable python."""
+        result = {
+            "start": self.default(obj.start),
+        }
+
+        if obj.duration != datetime.timedelta(seconds=0):
+            result["duration"] = create_iso8601_duration(obj.duration)
+
+        if obj.randomize_start != datetime.timedelta(seconds=0):
+            result["randomizeStart"] = create_iso8601_duration(obj.randomize_start)
+
+        return result
+
+    def serialize_interval(self, obj: Interval) -> dict[str, Any]:
+        """Serialize an Interval object to JSON encodable python."""
+        result = {
+            "id": obj.id,
+            "payloads": _serialize_values_map(obj.payloads),
+        }
+
+        if obj.interval_period:
+            result["intervalPeriod"] = self.default(obj.interval_period)
+
+        return result
+
+    def serialize_report_data(self, obj: ReportData) -> dict[str, Any]:
+        """Serialize a ReportData object to JSON encodable python."""
+        result = {
+            "resourceName": obj.resource_name,
+            "intervals": [self.default(interval) for interval in obj.intervals],
+        }
+
+        if obj.interval_period:
+            result["intervalPeriod"] = self.default(obj.interval_period)
+
+        return result
+
+    def serialize_report(self, obj: Report) -> dict[str, Any]:
+        """Serialize a Report object to JSON encodable python."""
+        result = {
+            "objectType": "REPORT",
+            "programID": obj.program_id,
+            "eventID": obj.event_id,
+            "clientName": obj.client_name,
+            "resources": [self.default(resource) for resource in obj.resources],
+        }
+
+        if obj.id is not None:
+            result["id"] = obj.id
+
+        if obj.created:
+            result["createdDateTime"] = self.default(obj.created)
+
+        if obj.modified:
+            result["modificationDateTime"] = self.default(obj.modified)
+
+        if obj.report_name:
+            result["reportName"] = obj.report_name
+
+        if obj.payload_descriptors:
+            result["payloadDescriptors"] = [
+                self.default(payload) for payload in obj.payload_descriptors
+            ]
+
+        return result
+
+    def serialize_event(self, obj: Event) -> dict[str, Any]:
```

Comment:
You could use a TypedDict here to get more hints in the IDE and have a full overview that is easy to compare with the docs. If you want to remove unset fields, you could do that more generically.